[v6] custom validation docs (#422)
Laravel-Backpack/docs: 6.x/custom-validation-rules.md

> # Custom Validation Rules

Reviewer: Ok so after going through all the rules, I think I've spotted some inconsistencies in the API, which make these difficult to understand/remember/use. Here's a rundown of the current API:

```php
'avatar'      => ValidUpload::fileRules()->attributeRules(),
'attachments' => ValidUploadMultiple::fileRules()->arrayRules(),
'gallery'     => ValidRepeatable::fieldRules()->itemRules([
                'title' => 'required',
                'image' => 'required'
            ]),
```

This is... all over the place. We need to find a way to name rule methods in a consistent way.

My questions are:
(1) In `ValidUpload` what's `attributeRules()`? I don't understand.
(2) Does `ValidUploadMultiple->arrayRules()` do the same as  `ValidRepeatable::fieldRules()`? In that case we need to find some common names.
(3) To make the API consistent across custom validation rules, what do you think about:
- (A) having `fieldRules()` on ALL custom validations, which validates the big field; all custom validation rules will need that part, right?
- (B) having `itemRules()` on ALL custom validators that have sub-items; the difference being, `ValidUploadMultiple::itemRules()` will validate the files (those are the items) and `ValidRepeatable::itemRules()` will validate the sub-items/entries (those are the items).

After the above, the API would become:
```php
'avatar'      => ValidUpload::fieldRules(),
'attachments' => ValidUploadMultiple::fieldRules()->itemRules(),
'gallery'     => ValidRepeatable::fieldRules()->itemRules([
                'title' => 'required',
                'image' => 'required'
            ]),
```

Does the above make sense and cover everything? Let me know.

Reviewer: TODO:
- [ ] add static `fieldRules()` to `ValidUpload`
- [ ] change docs to reflect recommended syntax:

```php
'avatar'      => ValidUpload::fieldRules()->fileRules(),
'attachments' => ValidUploadMultiple::fieldRules()->fileRules(), // alias: fileRules() == itemRules()
'gallery'     => ValidRepeatable::fieldRules()->itemRules([
                    'title' => 'required',
                    'image' => 'required',
                ]),
```

---

<a name="about"></a>
## About

> Along with the provided Laravel validation rules, we created a few custom validation rules that you can use to validate Backpack fields in an easier way.
> No more messing with `.*` and such on validation rules. Just define fluently what you want to validate.

Reviewer: ```suggestion
Some Backpack fields are more difficult to validate using standard Laravel validation rules. So we've created a few custom validation rules, that will make validation them dead-simple.
```

<a name="valid-upload-validation-rule"></a>
## `ValidUpload` for `upload` field type

The `ValidUpload` rule is used to validate the `upload` field type. Using the custom rule helps developer avoid to setting different validation rules for different operations.(create/update).

```php
// for the field
CRUD::field('avatar')->type('upload');

// you can write the validation rule as

> 'avatar' => ValidUpload::fileRules(File::types(['jpg','png'])->max(2048))->attributeRules('required'),

Reviewer: Wait... what's the API here? What is `fileRules()` and what is `attributeRules()`? There's a single file getting uploaded here soo... I don't get it...

```

The uploader will automatically handle the `sometimes` case for you and will only validate the field if it's present in the request.

<a name="valid-upload-multiple-validation-rule"></a>
## `ValidUploadMultiple` for `upload_multiple` field type

> The `ValidUploadMultiple` rule helps developer by having two sets of specific validation, one that applies to the "global array of files", and the "file validation" for the uploaded files.

Reviewer: ```suggestion
The `ValidUploadMultiple` rule helps you define two sets of validation rules, in one go:
- `arrayRules()` for the entire field (the array of entries);
- `fileRules()` for each uploaded file;
```

```php
// for the field
CRUD::field('attachments')->type('upload_multiple');

// you can write the validation rule as
'attachments' => ValidUploadMultiple::fileRules(File::types(['pdf'])->max(10000))
            ->arrayRules(['required_if:other_field,test_value', 'min:2', 'max:5'])
            ,    

```

<a name="valid-repeatable-validation-rule"></a>
## `ValidRepeatable` for `repeatable` field type

> The `ValidRepeatable` (works the same for relationships managed through the repeatable interface), is a validation rule that eases developer work by validating the repeatable fields in a more convinient way.

Reviewer: ```suggestion
The `ValidRepeatable` makes it easy to validate fields that have subfields (the `repeatable` field, but also the `relationship` field when you define subfields). You can define two sets of validation:
- `fieldRules()` to define rules on the field itself;
- `itemRules()` to define rules on the sub-items (rows);
```

<a name="valid-upload-validation-rule"></a>
## `ValidUpload` for `upload` field type

The `ValidUpload` rule is used to validate the `upload` field type. Using the custom rule helps developer avoid to setting different validation rules for different operations, (create/update).

```php
// for the field
CRUD::field('avatar')->type('upload');

// you can write the validation rule as

use Backpack\CRUD\app\Library\Validation\Rules\ValidUpload;

> 'avatar' => ValidUpload::field('required')->file(File::types(['jpg','png'])->max(2048)),

Reviewer: Doesn't this do the same, but look cleaner? Otherwise at first I thought you can do `max()` on the ValidUpload itself 🤦‍♂️ I think people are going to just _assume_ they can use class syntax for validation too, since it's a Laravel thing, not a Backpack thing: 

```suggestion
'avatar' => ValidUpload::field('required')->file('mimes:jpg,png|max:2048'),
```

```

> The `::field()` constructor accepts the rules for the field, while `->file()` accepts the specific rules for files sent in field.
> This helps the uploader to automatically handle the `sometimes` case for you.

Reviewer: ```suggestion
The `::field()` constructor accepts the rules for the field, while `->file()` accepts the specific rules for files sent in field. The validation rule handles the `sometimes` case for you.
```

<a name="valid-upload-multiple-validation-rule"></a>
## `ValidUploadMultiple` for `upload_multiple` field type

> This rule replaces the definitions of  `attribute` and `attribute.*` rules, plus handling the scenario of create vs. update.
> 
> The `::field()` constructor accepts the rules for the field, while `->file()` accepts the specific rules for files sent in field.

Reviewer: ```suggestion
You can use this validation rule to handle validation for you `upload_multiple` field - both for the Create and the Update operation in one go:
- use the `::field()` constructor to define the rules for the field;
- use the `->file()` method for rules specific to the files sent in field;
```

```php
// for the field
CRUD::field('attachments')->type('upload_multiple');

// you can write the validation rule as

use Backpack\CRUD\app\Library\Validation\Rules\ValidUploadMultiple;

> 'attachments' => ValidUploadMultiple::field(['required_if:other_field,test_value', 'min:2', 'max:5'])->file(File::types(['pdf'])->max(10000)),    

Reviewer: Let's simplify the example. We want to make this look simple, not complicated:

```suggestion
'attachments' => ValidUploadMultiple::field(['min:2', 'max:5'])->file('mimes:pdf|max:10000'),    
```
